consensus/tests/engine_v2_tests/helper.go

```diff
@@ -576,19 +582,16 @@ func CreateBlock(blockchain *BlockChain, chainConfig *params.ChainConfig, starti
 			Coinbase:   common.HexToAddress(blockCoinBase),
 			Extra:      extraInBytes,
 		}
-		if int64(blockNumber) == (chainConfig.XDPoS.V2.SwitchBlock.Int64() + 1) { // This is the first v2 block, we need to copy the last v1 epoch master node list and inject into v2 validators
+		if int64(blockNumber) == (chainConfig.XDPoS.V2.SwitchBlock.Int64() + 1) { // This is the first v2 block, we need v2 validators aka masternodes
 			// Get last master node list from last v1 block
-			lastv1Block := blockchain.GetBlockByNumber(chainConfig.XDPoS.V2.SwitchBlock.Uint64())
-			masternodesFromV1LastEpoch := decodeMasternodesFromHeaderExtra(lastv1Block.Header())
-			header.Validators = masternodesFromV1LastEpoch
-
+			header.Validators = masternodes
 		} else if roundNumber%int64(chainConfig.XDPoS.Epoch) == 0 {
-			// epoch switch blocks, copy the master node list and inject into v2 validators
-			// Get last master node list from last v1 block
-			lastv1Block := blockchain.GetBlockByNumber(chainConfig.XDPoS.V2.SwitchBlock.Uint64())
-			masternodesFromV1LastEpoch := decodeMasternodesFromHeaderExtra(lastv1Block.Header())
-			header.Validators = masternodesFromV1LastEpoch
-
+			// epoch switch blocks, we need v2 validators aka masternodes
+			header.Validators = masternodes
+		} else if (header.Number.Uint64()%chainConfig.XDPoS.Epoch) == (chainConfig.XDPoS.Epoch-chainConfig.XDPoS.Gap+1) || header.Number.Uint64() == 1 { // This is the gapPlusOneBlock
+			// gapPlusOneBlock, we need v2 validators aka masternodes as next epoch
+			header.NextValidators = masternodes
+			// Add penalties to gapPlusOneBlock
```

Comment:
@liam-lai please check this change

Reply:
your code is better, you read masternodes from genesis file